src/precomputed_quantities.jl

```diff
@@ -111,16 +134,218 @@ function ts(Yc, K, Φ, thermo_params, energy_form, moisture_model)
+                    q_ice⁰ =
+                        divide_by_ρa(ρaq_ice⁰, ρa⁰, Yc.ρq_ice, Yc.ρ, a_min)
+                    (; q_pt = TD.PhasePartition(q_tot⁰, q_liq⁰, q_ice⁰))
+                end
+            end
+            return thermo_state(
+                thermo_params;
+                p,
+                energy_var...,
+                moisture_var...,
+            )
+        end
+
+        for j in 1:n_mass_flux_subdomains(turbconv_model)
+            set_velocity_quantities!(
+                ᶜuʲs.:($j),
+                ᶠu³ʲs.:($j),
+                ᶜKʲs.:($j),
+                Y.f.sgsʲs.:($j).w,
+                Y.c,
+            )
+        end
+        @. ᶜtsʲs = tsʲs(Y.c, ᶜKʲs, ᶜp, ᶜΦ)
+
+        @. ᶜρa⁰ = Y.c.ρ - sum(sgsʲ -> sgsʲ.ρa, Y.c.sgsʲs)
+        Fields.bycolumn(axes(Y.c)) do colidx
+            @. ᶠw⁰[colidx] = w⁰(
+                ᶠinterp(ᶜρa⁰[colidx]),
+                ᶠinterp(map(sgsʲ -> sgsʲ.ρa, Y.c.sgsʲs[colidx])),
+                map(sgsʲ -> sgsʲ.w, Y.f.sgsʲs[colidx]),
+                ᶠinterp(Y.c.ρ[colidx]),
+                Y.f.w[colidx],
+            )
+        end
+        set_velocity_quantities!(ᶜu⁰, ᶠu³⁰, ᶜK⁰, ᶠw⁰, Y.c)
+        @. ᶜK⁰ += divide_by_ρa(Y.c.sgs⁰.ρatke, ᶜρa⁰, 0, Y.c.ρ, a_min)
+        @. ᶜts⁰ = ts⁰(Y.c, ᶜρa⁰, ᶜK⁰, ᶜp, ᶜΦ)
+    end
+end
+
+"""
+    diagnostic_edmfx_quantities(Y, p, t)
+
+Allocates, sets, and returns `ᶜsgs⁰`, `ᶜsgsᴰ`, `ᶜuᴰ`, and `ᶜtsᴰ` in a manner
+that is consistent with `set_precomputed_quantities!`. This function should only
+be used to generate diagnostics for `EDMFX`, and it should only be called after
+`set_precomputed_quantities!` has been called.
+"""
+function diagnostic_edmfx_quantities(Y, p, t)
+    (; energy_form, moisture_model, turbconv_model) = p.atmos
+    thermo_params = CAP.thermodynamics_params(p.params)
+    (; ᶜp, ᶜΦ) = p
+    ᶠinterp = Operators.InterpolateC2F(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
+
+    @assert turbconv_model isa EDMFX
+
+    (; a_min) = turbconv_model
+    (ᶠwᴰ, ᶜuᴰ, ᶠu³ᴰ, ᶜKᴰ) = similar.((p.ᶠw⁰, p.ᶜu⁰, p.ᶠu³⁰, p.ᶜK⁰))
+
+    wᴰ(ρaʲs, wʲs, ρ, w) =
+        divide_by_ρa(sum(ρaʲs .* wʲs), sum(ρaʲs), ρ * w, ρ, a_min)
+    sgsᴰ(Yc) = map(+, Yc.sgsʲs...)
+    sgs⁰(Yc) = merge(map(-, times_a(Yc, 1), sgsᴰ(Yc)), Yc.sgs⁰)
+
+    Fields.bycolumn(axes(Y.c)) do colidx
+        @. ᶠwᴰ[colidx] = wᴰ(
+            ᶠinterp(map(sgsʲ -> sgsʲ.ρa, Y.c.sgsʲs[colidx])),
+            map(sgsʲ -> sgsʲ.w, Y.f.sgsʲs[colidx]),
+            ᶠinterp(Y.c.ρ[colidx]),
+            Y.f.w[colidx],
+        )
+    end
+    set_velocity_quantities!(ᶜuᴰ, ᶠu³ᴰ, ᶜKᴰ, ᶠwᴰ, Y.c)
+
+    ᶜsgsᴰ = @. sgsᴰ(Y.c)
+    ᶜsgs⁰ = @. sgs⁰(Y.c)
+    ᶜtsᴰ = @. tsʲ(
+        ᶜsgsᴰ,
+        Y.c,
+        ᶜKᴰ,
+        ᶜp,
+        ᶜΦ,
+        a_min,
+        thermo_params,
+        energy_form,
+        moisture_model,
+    )
+
+    return (; ᶜsgs⁰, ᶜsgsᴰ, ᶜuᴰ, ᶜtsᴰ)
```

Comment:
`ᶜsgs⁰`,  `ᶜsgsᴰ` are very hard to tell apart. I think we should do something else unfortunately

Reply:
Oh, I agree with this. subscript 0 always looks like \theta to me.

Reply:
Per an offline discussion, I'll change the ᴰ to a ⁺. That should be easier to distinguish, and it will match the ⟨⋅⟩₊ notation from Tapio's Overleaf.